Serialization issues in unit tests on OSX

When building with unit tests on OSX, I encounter the following linker errors:
```
Undefined symbols for architecture x86_64:
  "void El::load<cereal::BinaryInputArchive, float, 1>(cereal::BinaryInputArchive&, El::Matrix<float, (hydrogen::Device)0>&)", referenced from:
      void ____C_A_T_C_H____T_E_M_P_L_A_T_E____T_E_S_T____F_U_N_C____0<El::Matrix<float, (hydrogen::Device)0> >() in serialize_matrix_test.cpp.o
      std::__1::enable_if<std::is_polymorphic<El::Matrix<float, (hydrogen::Device)0> >::value, void>::type cereal::load<cereal::BinaryInputArchive, El::Matrix<float, (hydrogen::Device)0>, std::__1::default_delete<El::Matrix<float, (hydrogen::Device)0> > >(cereal::BinaryInputArchive&, std::__1::unique_ptr<El::Matrix<float, (hydrogen::Device)0>, std::__1::default_delete<El::Matrix<float, (hydrogen::Device)0> > >&) in serialize_matrix_test.cpp.o
  "void El::save<cereal::BinaryOutputArchive, float, (hydrogen::Device)0, 1>(cereal::BinaryOutputArchive&, El::Matrix<float, (hydrogen::Device)0> const&)", referenced from:
      void ____C_A_T_C_H____T_E_M_P_L_A_T_E____T_E_S_T____F_U_N_C____0<El::Matrix<float, (hydrogen::Device)0> >() in serialize_matrix_test.cpp.o
      std::__1::enable_if<(std::is_polymorphic<El::Matrix<float, (hydrogen::Device)0> >::value) && (!(std::is_abstract<El::Matrix<float, (hydrogen::Device)0> >::value)), void>::type cereal::save<cereal::BinaryOutputArchive, El::Matrix<float, (hydrogen::Device)0>, std::__1::default_delete<El::Matrix<float, (hydrogen::Device)0> > >(cereal::BinaryOutputArchive&, std::__1::unique_ptr<El::Matrix<float, (hydrogen::Device)0>, std::__1::default_delete<El::Matrix<float, (hydrogen::Device)0> > > const&) in serialize_matrix_test.cpp.o
ld: symbol(s) not found for architecture x86_64
```
I think this will be straightforward to resolve with some additional ETI, but I haven't dug into it yet.

--- a/src/utils/serialize_matrices/AbstractDistMatrix.cpp
+++ b/src/utils/serialize_matrices/AbstractDistMatrix.cpp
@@ -1,0 +1,78 @@
+////////////////////////////////////////////////////////////////////////////////
+// Copyright (c) 2014-2019, Lawrence Livermore National Security, LLC.
+// Produced at the Lawrence Livermore National Laboratory.
+// Written by the LBANN Research Team (B. Van Essen, et al.) listed in
+// the CONTRIBUTORS file. <lbann-dev@llnl.gov>
+//
+// LLNL-CODE-697807.
+// All rights reserved.
+//
+// This file is part of LBANN: Livermore Big Artificial Neural Network
+// Toolkit. For details, see http://software.llnl.gov/LBANN or
+// https://github.com/LLNL/LBANN.
+//
+// Licensed under the Apache License, Version 2.0 (the "Licensee"); you
+// may not use this file except in compliance with the License.  You may
+// obtain a copy of the License at:
+//
+// http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or
+// implied. See the License for the specific language governing
+// permissions and limitations under the license.
+////////////////////////////////////////////////////////////////////////////////
+
+#include "common.hpp"
+#include <lbann/utils/serialization/serialize_matrices.hpp>
+#include <lbann/utils/serialization/serialize_matrices_impl.hpp>
+
+#ifdef LBANN_HAS_CEREAL_BINARY_ARCHIVES
+#define ETI_BINARY_SAVELOAD(TYPE)                                      \
+  template void El::load(cereal::BinaryInputArchive&,                          \
+                         El::AbstractDistMatrix<TYPE>&);                       \
+  template void El::save(cereal::BinaryOutputArchive&,                         \
+                         El::AbstractDistMatrix<TYPE> const&);
+#else
+#define ETI_BINARY_SAVELOAD(TYPE)
+#endif // LBANN_HAS_CEREAL_BINARY_ARCHIVES
+
+#ifdef LBANN_HAS_CEREAL_JSON_ARCHIVES // Not yet supported
+#define ETI_JSON_SAVELOAD(TYPE)                                        \
+  template void El::load(cereal::JSONInputArchive&,                            \
+                         El::AbstractDistMatrix<TYPE>&);                       \
+  template void El::save(cereal::JSONOutputArchive&,                           \
+                         El::AbstractDistMatrix<TYPE> const&);
+#else
+#define ETI_JSON_SAVELOAD(TYPE)
+#endif // LBANN_HAS_CEREAL_JSON_ARCHIVES
+
+#ifdef LBANN_HAS_CEREAL_PORTABLE_BINARY_ARCHIVES // Not yet supported
+#define ETI_PORTABLEBINARY_SAVELOAD(TYPE)                              \
+  template void El::load(cereal::PortableBinaryInputArchive&,                  \
+                         El::AbstractDistMatrix<TYPE>&);                       \
+  template void El::save(cereal::PortableBinaryOutputArchive&,                 \
+                         El::AbstractDistMatrix<TYPE> const&);
+#else
+#define ETI_PORTABLEBINARY_SAVELOAD(TYPE)
+#endif // LBANN_HAS_CEREAL_PORTABLE_BINARY_ARCHIVES
+
+#ifdef LBANN_HAS_CEREAL_XML_ARCHIVES
+#define ETI_XML_SAVELOAD(TYPE)                                         \
+  template void El::load(cereal::XMLInputArchive&,                             \
+                         El::AbstractDistMatrix<TYPE>&);                       \
+  template void El::save(cereal::XMLOutputArchive&,                            \
+                         El::AbstractDistMatrix<TYPE> const&);
+#else
+#define ETI_XML_SAVELOAD(TYPE)
+#endif // LBANN_HAS_CEREAL_XML_ARCHIVES
+
+// Enumerate all the valid data types.
+#define PROTO(T)                                                               \
+  ETI_BINARY_SAVELOAD(T)                                                       \
+  ETI_JSON_SAVELOAD(T)                                                         \
+  ETI_PORTABLEBINARY_SAVELOAD(T)                                               \
+  ETI_XML_SAVELOAD(T)
+
+#include <lbann/macros/instantiate.hpp>

--- a/src/utils/serialize_matrices/CMakeLists.txt
+++ b/src/utils/serialize_matrices/CMakeLists.txt
@@ -1,5 +1,6 @@
 # Add the source files for this directory
 set_full_path(THIS_DIR_SOURCES
+  AbstractDistMatrix.cpp
   CIRC_CIRC.cpp
   MC_MR.cpp
   MC_STAR.cpp

--- a/src/utils/serialize_matrices/Matrix.cpp
+++ b/src/utils/serialize_matrices/Matrix.cpp
@@ -30,29 +30,87 @@
 // Register types outside Cereal's namespace.
 #define LBANN_COMMA ,
 
+#ifdef LBANN_HAS_CEREAL_BINARY_ARCHIVES
+#define ETI_BINARY_SAVELOAD(TYPE, DEVICE)                                      \
+  template void El::load(cereal::BinaryInputArchive&,                          \
+                         El::Matrix<TYPE, El::Device::DEVICE>&);               \
+  template void El::save(cereal::BinaryOutputArchive&,                         \
+                         El::Matrix<TYPE, El::Device::DEVICE> const&);         \
+  template void El::load(lbann::RootedBinaryInputArchive&,                     \
+                         El::Matrix<TYPE, El::Device::DEVICE>&);               \
+  template void El::save(lbann::RootedBinaryOutputArchive&,                    \
+                         El::Matrix<TYPE, El::Device::DEVICE> const&);
+#else
+#define ETI_BINARY_SAVELOAD(TYPE, DEVICE)
+#endif // LBANN_HAS_CEREAL_BINARY_ARCHIVES
+
+#ifdef LBANN_HAS_CEREAL_JSON_ARCHIVES // Not yet supported
+#define ETI_JSON_SAVELOAD(TYPE, DEVICE)                                        \
+  template void El::load(cereal::JSONInputArchive&,                            \
+                         El::Matrix<TYPE, El::Device::DEVICE>&);               \
+  template void El::save(cereal::JSONOutputArchive&,                           \
+                         El::Matrix<TYPE, El::Device::DEVICE> const&);         \
+  template void El::load(lbann::RootedJSONInputArchive&,                       \
+                         El::Matrix<TYPE, El::Device::DEVICE>&);               \
+  template void El::save(lbann::RootedJSONOutputArchive&,                      \
+                         El::Matrix<TYPE, El::Device::DEVICE> const&);
+#else
+#define ETI_JSON_SAVELOAD(TYPE, DEVICE)
+#endif // LBANN_HAS_CEREAL_JSON_ARCHIVES
+
+#ifdef LBANN_HAS_CEREAL_PORTABLE_BINARY_ARCHIVES // Not yet supported
+#define ETI_PORTABLEBINARY_SAVELOAD(TYPE, DEVICE)                              \
+  template void El::load(cereal::PortableBinaryInputArchive&,                  \
+                         El::Matrix<TYPE, El::Device::DEVICE>&);               \
+  template void El::save(cereal::PortableBinaryOutputArchive&,                 \
+                         El::Matrix<TYPE, El::Device::DEVICE> const&);         \
+  template void El::load(lbann::RootedPortableBinaryInputArchive&,             \
+                         El::Matrix<TYPE, El::Device::DEVICE>&);               \
+  template void El::save(lbann::RootedPortableBinaryOutputArchive&,            \
+                         El::Matrix<TYPE, El::Device::DEVICE> const&);
+#else
+#define ETI_PORTABLEBINARY_SAVELOAD(TYPE, DEVICE)
+#endif // LBANN_HAS_CEREAL_PORTABLE_BINARY_ARCHIVES
+
+#ifdef LBANN_HAS_CEREAL_XML_ARCHIVES
+#define ETI_XML_SAVELOAD(TYPE, DEVICE)                                         \
+  template void El::load(cereal::XMLInputArchive&,                             \
+                         El::Matrix<TYPE, El::Device::DEVICE>&);               \
+  template void El::save(cereal::XMLOutputArchive&,                            \
+                         El::Matrix<TYPE, El::Device::DEVICE> const&);         \
+  template void El::load(lbann::RootedXMLInputArchive&,                        \
+                         El::Matrix<TYPE, El::Device::DEVICE>&);               \
+  template void El::save(lbann::RootedXMLOutputArchive&,                       \
+                         El::Matrix<TYPE, El::Device::DEVICE> const&);
+#else
+#define ETI_XML_SAVELOAD(TYPE, DEVICE)
+#endif // LBANN_HAS_CEREAL_XML_ARCHIVES
+
 // Register the local Matrix classes
-#define REGISTER_LOCAL_MATRIX(TYPE, DEVICE)             \
-  CEREAL_REGISTER_TYPE_WITH_NAME(                       \
-    El::Matrix<TYPE LBANN_COMMA El::Device::DEVICE>,    \
-    "Matrix(" #TYPE "," #DEVICE ")")                    \
-  CEREAL_REGISTER_POLYMORPHIC_RELATION(                 \
-    El::AbstractMatrix<TYPE>,                           \
-    El::Matrix<TYPE LBANN_COMMA El::Device::DEVICE>)
+#define REGISTER_LOCAL_MATRIX(TYPE, DEVICE)                                    \
+  CEREAL_REGISTER_TYPE_WITH_NAME(                                              \
+    El::Matrix<TYPE LBANN_COMMA El::Device::DEVICE>,                           \
+    "Matrix(" #TYPE "," #DEVICE ")")                                           \
+  CEREAL_REGISTER_POLYMORPHIC_RELATION(                                        \
+    El::AbstractMatrix<TYPE>,                                                  \
+    El::Matrix<TYPE LBANN_COMMA El::Device::DEVICE>)                           \
+  ETI_BINARY_SAVELOAD(TYPE, DEVICE)                                            \
+  ETI_JSON_SAVELOAD(TYPE, DEVICE)                                              \
+  ETI_PORTABLEBINARY_SAVELOAD(TYPE, DEVICE)                                    \
+  ETI_XML_SAVELOAD(TYPE, DEVICE)
 
 // Enumerate the valid device types. This is done here rather than
 // "instantiate_device.hpp" because it's a miniscule overhead and the
 // slightly prettier printing is worth it to me -- the namespace and
 // enum class are irrelevant to users!
 #if defined LBANN_HAS_GPU
-#define REGISTER_ALL_MATRIX_DEVICES(TYPE)   \
-  REGISTER_LOCAL_MATRIX(TYPE, CPU)          \
+#define REGISTER_ALL_MATRIX_DEVICES(TYPE)                                      \
+  REGISTER_LOCAL_MATRIX(TYPE, CPU)                                             \
   REGISTER_LOCAL_MATRIX(TYPE, GPU)
 #else
-#define REGISTER_ALL_MATRIX_DEVICES(TYPE)       \
-  REGISTER_LOCAL_MATRIX(TYPE, CPU)
+#define REGISTER_ALL_MATRIX_DEVICES(TYPE) REGISTER_LOCAL_MATRIX(TYPE, CPU)
 #endif // defined LBANN_HAS_GPU
 
 // Enumerate all the valid data types.
-#define PROTO(T)                                \
-  REGISTER_ALL_MATRIX_DEVICES(T)
+#define PROTO(T) REGISTER_ALL_MATRIX_DEVICES(T)
 #include <lbann/macros/instantiate.hpp>

